src/rdds/lib/process_pool/process_pool.py

```diff
@@ -0,0 +1,235 @@
+
+
+@dataclass
+class DummyProcess:
+    name: str
+    exitcode: int
+
+
+@dataclass
+class Task:
+    args: Iterable
+    process: Union[Process, DummyProcess]
+
+
+CompletedTaskQueue = Queue
+
+
+class ProcessPool:
+
+    """
+    Implements a simple multiprocessing pool supporting synchronous and asynchronous processing mode.
+    """
+
+    def __init__(self,
+                 function: Callable,
+                 args: Iterable,
+                 process_names: List[str] = None,
+                 workers: int = cpu_count() - 1  # Reserve a core for main process
+                 ):
+        """
+        :param function: The function to execute concurrently
+        :param args: Function arguments
+        :param process_names: List of names to assign to processes for traceability
+        :param workers: Maximum concurrent workers running 'fn' simultaneously
+        """
+        self._function = function
+        self._args = args
+        self._nr_expected_tasks: int = len(self._args)
+        self._process_names = process_names if process_names else [None] * self._nr_expected_tasks
+        self._workers: int = workers
+        self._running_tasks: List[Task] = []
+        self._running_tasks_lock: Lock = Lock()
+        self._completed_tasks: List[Task] = []
+        self._dispatch_thread: Thread = None
+        self._collect_stopped_tasks_thread: Thread = None
+        self._completed_tasks_async_queue = CompletedTaskQueue()
+        self._processing_complete: Event = Event()
+        self._ctx = self.get_context()
+
+    @staticmethod
+    def get_context() -> SpawnContext:
+        """
+        Return context for processes, queues etc.
+        Don't fork py process in pool, this reduces RAM consumption by not inheriting objects.
+        Add top of file before running any multiprocessing-module related code (breaks otherwise).
+        :return:
+        """
+        return get_context('spawn')
+
+    @property
+    def nr_expected_tasks(self) -> int:
+        """
+        Return the number of expected tasks to be executed by the pool.
+        :return:
+        """
+        return self._nr_expected_tasks
+
+    def _dispatch(self) -> None:
+        """
+        Method that dispatches processes. Thread exits after all tasks has been dispatched.
+        """
+
+        def wait_until_available_worker() -> None:
+            # TODO: This thread will consume lots of CPU if spinning on IF condition
+            while True:
+                # If there's an available slot for a new process, break out
+                self._running_tasks_lock.acquire()
+                if len(self._running_tasks) < self._workers:
+                    self._running_tasks_lock.release()
+                    break
+                self._running_tasks_lock.release()
+
+        def dispatch_new_process() -> None:
+            self._running_tasks_lock.acquire()
+            process: Process = self._ctx.Process(target=self._function,
```

Comment:
```suggestion
            process = self._ctx.Process(target=self._function,
```
No need to type hint when calling a constructor. It is already clear what the return will be.